### PR TITLE
add toleration to run on masters

### PR DESCRIPTION
**Pull request by vxav**
This PR:

- adds toleration and nodeselector to run controller on masters.
Towards https://github.com/giantswarm/giantswarm/issues/24885

### Checklist

- [x] Update changelog in CHANGELOG.md.
- [x] Make sure `values.yaml` and `values.schema.json` are valid.
- [x] Update lastpass values if required.

### CODE REVIEW

**Comment by erkanerol** [comment]
@vxav can we talk about this in a meeting like kaas sync? we need to do this for many apps and I am not sure how we should solve this issue.

File: config/kustomize/patches/deployment-resources.yaml [hunk 1]
```diff
@@ -7,6 +7,12 @@ metadata:
 spec:
   template:
     spec:
+      nodeSelector:
+        node-role.kubernetes.io/master: ""
+      tolerations:
+        - key: "node-role.kubernetes.io/master"
```
Comment (erkanerol) [inline config/kustomize/patches/deployment-resources.yaml:13]: `master` is the old name. the new one should be something like `control-plane`

**Comment by jkremser** [comment]
what about reusing one of the existing priority classes for this?

```
k get priorityclasses.scheduling.k8s.io
NAME                      VALUE        GLOBAL-DEFAULT   AGE
giantswarm-critical       1000000000   false            10d
prometheus                500000000    false            10d
system-cluster-critical   2000000000   false            10d
system-node-critical      2000001000   false            10d
```

```
k get pods --all-namespaces -o custom-columns=NAME:.metadata.name,PRIORITY:.spec.priorityClassName,VALUE:.spec.priority --sort-by=.spec.priority

...
prometheus-guppy-0                                                prometheus                500000000
kyverno-74c449ffff-mmr99                                          giantswarm-critical       1000000000
kyverno-74c449ffff-jc4hl                                          giantswarm-critical       1000000000
etcd-backup-operator-64b465f7bb-2zlxs                             giantswarm-critical       1000000000
metrics-server-6c68b7d978-cndkv                                   giantswarm-critical       1000000000
cert-exporter-deployment-597dc7c698-sh9nx                         giantswarm-critical       1000000000
metrics-server-6c68b7d978-wz5j2                                   giantswarm-critical       1000000000
chart-operator-c586bbb76-zgk6c                                    giantswarm-critical       1000000000
kyverno-74c449ffff-94qc6                                          giantswarm-critical       1000000000
nginx-ingress-controller-app-7686d87649-c8dtx                     system-cluster-critical   2000000000
coredns-workers-7594f8bffd-gjlww                                  system-cluster-critical   2000000000
coredns-workers-7594f8bffd-sfd4j                                  system-cluster-critical   2000000000
kube-state-metrics-698f6b5d8c-84r55                               system-cluster-critical   2000000000
nginx-ingress-controller-app-7686d87649-bbjrm                     system-cluster-critical   2000000000
cilium-operator-58fc58976d-pglmp                                  system-cluster-critical   2000000000
cilium-operator-58fc58976d-lqwrx                                  system-cluster-critical   2000000000
kube-scheduler-guppy-8fb68                                        system-node-critical      2000001000
...
```

`system-cluster-critical` could be a good fit

File: config/kustomize/patches/deployment-resources.yaml (hunk 1 above)
Comment (kopiczko) [reply config/kustomize/patches/deployment-resources.yaml:13]: `node-role.kubernetes.io/control-plane`

**Comment by kopiczko** [comment]
Isn't priority classes a different concept?

**Comment by jkremser** [comment]
> Isn't priority classes a different concept?

it is. But I assume the reason for all this is that we would like to somehow guarantee that the infra provider is always there, right? In which case the priority classes might be a better fit for this. Because forcing it to run on control-plane via node selector may fail if the resources on control-planes are low. Actually both concepts can be used together I think. Having the toleration there (but not the nodeSelector) + priorityClass combo. So that it's schedulable on every node + if the masters are low on resources, it will be scheduled on workers.

File: config/kustomize/patches/deployment-resources.yaml (hunk 1 above)
Comment (vxav) [reply config/kustomize/patches/deployment-resources.yaml]: Changed in https://github.com/giantswarm/cluster-api-provider-cloud-director-app/pull/30/commits/1dd91d022c77e55268a1cf6109826fad43fcd137

**Comment by vxav** [comment]
What about `nodeAffinity` with `preferredDuringSchedulingIgnoredDuringExecution`. Might be clearer that priorityClass wdyt?

**Comment by jkremser** [comment]
right, that weak/soft form of node affinity is probably better than setting some global magic number

**Review by erkanerol — APPROVED** [review]
LGTM

**Comment by vxav** [comment]
nodeAffinity configured in https://github.com/giantswarm/cluster-api-provider-cloud-director-app/pull/30/commits/ad74a432340ef3f4da59eac52e546f99c3b177a4. 
However, the maintainers set it to [required in the CPI](https://github.com/vmware/cloud-provider-for-cloud-director/blob/main/manifests/cloud-director-ccm.yaml#L179) and there's no node selection in the CSI. 🤷

**Comment by kopiczko** [comment]
> > Isn't priority classes a different concept?
> 
> it is. But I assume the reason for all this is that we would like to somehow guarantee that the infra provider is always there, right? In which case the priority classes might be a better fit for this.

I thought the problem is when there are not healthy worker nodes. In such scenario it's impossible to create a new worker node if we don't add the toleration. Priority class is a separate issue, but I agree combined with toleration may increase the chances to help.

I don't understand how affinity is supposed to help. Maybe it will make the initial capacity planning easier, but since it's soft affinity I'm not sure it will.

I.e. for me only two of the mentioned things make sense:

- toleration (required)
- priority class (optional)

**Comment by vxav** [comment]
My understanding is that using a soft affinity rule will make it run the controllers on the CP nodes as much as possible and use the workers if not possible ?

**Comment by kopiczko** [comment]
> My understanding is that using a soft affinity rule will make it run the controllers on the CP nodes as much as possible and use the workers if not possible ?

Yes. What problem does that solve?

**Comment by erkanerol** [comment]
What I believe is that we should use PriorityClass and Toleration. Soft affinity is optional. We shouldn't use NodeSelector or hard affinity. 

***Background***
Since management clusters manage themselves, CAPI/CAPO controllers are critical when there is a problem in some nodes or when we need to scale up nodes because of high demand. Now, the scheduler doesn't schedule these controllers when there is not enough resource. When the controllers don't run, we cannot scale up the nodes. In the end, we need to kill some pods to find enough space in worker nodes for the controllers. Since these controllers are critical for the cluster itself, we should behave in a way that these are part of the control plane. At least softly. 

***Scenarios***
- Everything is fine: Then let's schedule the controllers to CP nodes. 
- Everything is fine but there is a shortage in CP nodes: Thanks to soft affinity, scheduler will schedule the controllers on worker nodes. There is no negative side-effect. If there is enough space in worker nodes, we will not need to scale up CP nodes. 
- The controller pods restarted somehow when there is not enough space in the cluster: Some pending pods can start during this period. Then scheduler cannot find a place to schedule the controllers. Then it will kill some non-critical pods to make room for the controllers thanks to PriorityClass. 
- Worker nodes became non-functional when the controllers were running on worker nodes: Thanks to toleration&soft affinity, scheduler can reschedule the controllers to CP nodes. The controllers must run eventually but we don't need %100 uptime. Rescheduling is OK.
- Worker nodes became non-functional when the controllers were running on worker nodes and there is no enough space in CP nodes: Thanks to PriorityClass, scheduler will kill the other non-critical pods in CP nodes and we will have the controllers up&running on CP nodes.

**Comment by vxav** [comment]
I agree with above opinions that `priorityClass` is a better option than both `nodeSelector` and `nodeAffinity`.

Switched to configurable `priorityClass` in https://github.com/giantswarm/cluster-api-provider-cloud-director-app/pull/30/commits/6105c96ec70202398a19e8d4f21eccb5c22a9a1a

File: helm/cluster-api-provider-cloud-director/values.yaml [hunk 2]
```diff
@@ -19,3 +19,5 @@ resources:
     memory: 512Mi
 
 skipRDE: true
+
+priorityClass: giantswarm-critical
```
Comment (kopiczko) [inline helm/cluster-api-provider-cloud-director/values.yaml:23]: To make it usable from outside Giant Swarm we should set it to an empty string. Add `if` statement if the value is empty in the deployment and set `priorityClass` globally for cluster-api-provider-cloud-director-app in `config` repo.

**Review by kopiczko — APPROVED** [review]

**Pull request merged**